src/sparse.jl

```diff
@@ -0,0 +1,164 @@
+## Sparsity detector
+
+"""
+    AbstractSparsityDetector
+
+Abstract supertype for sparsity pattern detectors, defined for instance in [Symbolics.jl](https://github.com/JuliaSymbolics/Symbolics.jl).
+
+# Required methods
+
+- [`jacobian_sparsity`](@ref)
+- [`hessian_sparsity`](@ref)
+"""
+abstract type AbstractSparsityDetector end
+
+"""
+    jacobian_sparsity(f, x, sd::AbstractSparsityDetector)::AbstractMatrix{Bool}
+    jacobian_sparsity(f!, y, x, sd::AbstractSparsityDetector)::AbstractMatrix{Bool}
+
+Use `sd` to construct a (typically sparse) matrix `S` describing the pattern of nonzeroes in the Jacobian of `f` (resp. `f!`) applied at `x` (resp. `(y, x)`).
+"""
+function jacobian_sparsity end
+
+"""
+    hessian_sparsity(f, x, sd::AbstractSparsityDetector)::AbstractMatrix{Bool}
+
+Use `sd` to construct a (typically sparse) matrix `S` describing the pattern of nonzeroes in the Hessian of `f` applied at `x`.
+"""
+function hessian_sparsity end
+
+"""
+    NoSparsityDetector <: AbstractSparsityDetector
+
+Trivial sparsity detector, which always returns a full sparsity pattern (only ones, no zeroes).
+
+# See also
+
+- [`AbstractSparsityDetector`](@ref)
+"""
+struct NoSparsityDetector <: AbstractSparsityDetector end
+
+jacobian_sparsity(f, x, ::NoSparsityDetector) = ones(Bool, length(f(x)), length(x))
+jacobian_sparsity(f!, y, x, ::NoSparsityDetector) = ones(Bool, length(y), length(x))
+hessian_sparsity(f, x, ::NoSparsityDetector) = ones(Bool, length(x), length(x))
+
+## Coloring algorithm
+
+"""
+    AbstractColoringAlgorithm
+
+Abstract supertype for Jacobian/Hessian coloring algorithms, defined for example in [SparseDiffTools.jl](https://github.com/JuliaDiff/SparseDiffTools.jl).
+
+# Required methods
+
+- [`column_coloring`](@ref)
+- [`row_coloring`](@ref)
+"""
+abstract type AbstractColoringAlgorithm end
```

Comment:
What's the plan for de-conflicting it with https://github.com/JuliaArrays/ArrayInterface.jl/blob/master/src/ArrayInterface.jl#L367 ?

Reply:
Yeah I had seen that, but it is a conscious proposition from me. Since ArrayInterface has way more dependencies than ADTypes, I didn't want to import the former here because ADTypes is gonna be at the root of hopefully the whole ecosystem.
Two solutions:
- (my least favorite) import ADTypes in ArrayInterface instead
- (my favorite) keep this code in ADTypes and remove it from ArrayInterface altogether, since it is very specific to differentiation

Reply:
But then who defines the matrix coloring function definitions for base structured types?

Reply:
Yeah I hadn't thought of that. Then I guess it's better if those names live in ADTypes and are imported by ArrayInterface? I don't know if matrix coloring has other prominent uses besides autodiff

Reply:
It's mainly for autodiff.

Reply:
okay then it makes sense to put it here